### PR TITLE
Improved and clearer Java installation tutorial with screenshots

### PR DESCRIPTION
**Pull request by lordofpipes**
I spent some time making the tutorial for installing Java on par with the tutorial that we use for our server. It's now relevant to modern Minecraft with its new Java 17 requirements. It also avoids recommending Oracle JRE which could be a licensing trap for some users ever since the 2019 changes.

This also fixes the table formatting on the `Pixel Format Not Accelerated` article. Previously the tables were all squished together.

There's no license on this repository but feel free to consider this particular contribution as CC0 / public domain. Feel free to copy it verbatim to your website if you're a server or a hosting company.

Related discussion:

- Solves #82 
- #106

### CODE REVIEW

**Comment by randomairborne** [comment]
This is perhaps a bit too comprehensive- almost feel like there should be links to the sections that will be needed the most, x86 Windows and Linux, or perhaps they should be at the top.

**Comment by lordofpipes** [comment]
> This is perhaps a bit too comprehensive- almost feel like there should be links to the sections that will be needed the most, x86 Windows and Linux, or perhaps they should be at the top.

How about bringing "Possible symptoms of incorrect Java version" and "Verifying if Java is installed" to the bottom of the page and adding some anchor links to the top? The symptoms section is mainly for search engine accessibility anyways.

**Comment by lordofpipes** [comment]
With a bit of Javascript, the quick links could also function to automatically open up the spoilers.

**Review by Southpaw1496 — CHANGES_REQUESTED** [review]
Hi

The changes I asked for are quite substantial, and I appreciate the work you've done that you'll now need to do over. To avoid this in future, you can either discuss this in a GitHub issue or on Discord (Discord will probably be more visible), or you can open draft pull requests as you work on contributions, so that we can see their progress and comment on them before they're done.

Also, since you're going to do most of your screenshots again, I'm going to tell you now that most of them are not very suitable. You've got too many markings and the screenshots are cluttered, especially in focused screenshots such as `adoptium-mac/openjdk-installer-2`. When markings are helpful, please use boxes where you can instead of circles and massive arrows (see: `adoptium-windows/openjdk-installer-4.jpg`).

Overall, though, the article is a well overdue rework, and I really appreciate the time that has gone into it :).

File: _help/installing-java/index.markdown [hunk 1]
```diff
@@ -5,71 +5,458 @@ name: "installing-java"
...
 
-## [Installing Java for Linux](#install-linux) {#install-linux}
+On Linux, your package manager will choose the appropriate version automatically. But if you are going with a third party Java distribution on Linux (such as Microsoft, Adoptium, or Azul) you can check manually by running `uname -m` in a terminal.
 
-#### Ubuntu/Most Debian distros
-- Open up your terminal
-- Enter in Terminal: `sudo apt update;sudo apt install openjdk-8-jre`, Press enter then enter your password.
-- After installing run `sudo update-alternatives --install /usr/bin/java java /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java 100;sudo update-alternatives --config java`
-- When it prompts you to, select jre 8 using your arrow keys
+### I am on Windows and I need latest Java (Version 17) {#install-windows}
```
[the remark this answers is not in the document]
Comment (lordofpipes) [reply _help/installing-java/index.markdown:56]: The original reason for the split is that I'm recommending Microsoft OpenJDK if you just need java 17, 64 bit, (the theory being that this is least chance of unexpected problems since it is an identical binary to what's used in Minecraft &mdash; has the same hash) and Adoptium if you need something older or 32 bit. This is not a great experience, because the Microsoft website is slightly harder to use because it does not auto detect platform for macOS, and of course *only* provides x64 and ARM.

If this article is to only recommend Adoptium, then yeah, this split is not necessary. I'll reorganize it accordingly.

File: _help/installing-java/index.markdown [hunk 2]
```diff
@@ -5,71 +5,458 @@ name: "installing-java"
...
-### Linux
-- Open Terminal (Ctrl + Alt + T on Ubuntu)
-- Enter in Terminal: `java -version`
-- If the output does not contain something like `openjdk version "1.8.0_ABC"` or if it returns `java: command not found` then Java is not installed; see the Linux section below for how to install Java.
-- If the output does not contain Java version 1.8 then it's highly recommended you reinstall Java; see the Linux section below for how to reinstall Java.
+The Minecraft launchers for Windows, macOS, and Linux ship with a builtin Java that the launcher automatically keeps up to date. This includes the Windows 7 and Windows 8.1 version of the launcher.
 
+If you are using the official launcher, normally you do not have to install Java on your computer unless you need it for applications other than Minecraft. If you are not are not sure if you are using the new Minecraft launcher or are getting "This application requires a Java Runtime Environment" when launching Minecraft, then you can download the Minecraft Launcher from [https://minecraft.net/download](https://minecraft.net/download).
 
+**TL;DR: Most people running vanilla Minecraft with the default launcher will not need this tutorial!**
```
[the remark this answers is not in the document]
Comment (lordofpipes) [reply _help/installing-java/index.markdown:27]: Some may arrive at this page via search engines &mdash; how about

> TL;DR: Most people running vanilla Minecraft with the default launcher will not need this tutorial, but if it has been suggested to you as a workaround, then continue.

**Comment by lordofpipes** [comment]
> The changes I asked for are quite substantial, and I appreciate the work you've done that you'll now need to do over. To avoid this in future, you can either discuss this in a GitHub issue or on Discord (Discord will probably be more visible), or you can open draft pull requests as you work on contributions, so that we can see their progress and comment on them before they're done.

To be clear, this is a tutorial mostly adapted from what I use for another community, so I'm not miffed about the need for changes for it to be suitable for the Hopper, since I had already written most of it without Hopper in mind. I'll work on these points, ty for the feedback.

File: _help/installing-java/index.markdown (hunk 1 above)
[the remark this answers is not in the document]
Comment (lordofpipes) [reply _help/installing-java/index.markdown]: Done.

**Comment by lordofpipes** [comment]
I have switched the tutorial to Adoptium-only and did all these changes you suggested. Changed all of the ellipses to rectangles and only used markings where it might be non-obvious.

Preview of the page on first load:

everything-collapsed

Preview of the page after expanding the spoilers:

everything-expanded

File: _help/installing-java/index.markdown [hunk 3]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
+```
+
+![Terminal with Java alternatives command executed](/static/images/help/installing-java/linux/install-java-4.jpg)
+
+" | markdownify }} </details>
+
+
+## Next steps
+
+Once you have installed the proper Java version, you should configure this in your launcher and give the game a go! Depending on your launcher you may need to manually navigate to the path of Adoptium OpenJDK.
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:241]: We actually have an article on this, https://minecrafthopper.net/help/guides/changing-java-version/

It's very, very outdated though, so I will work on updating it myself, and it can be merged before this article.

We don't need to deal with 3rd-party launchers either, you can get rid of this entire section.

File: _help/installing-java/index.markdown [hunk 4]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
+
+Run the sanity check to ensure you did everything correctly.
+
+![Java sanity check in MultiMC](/static/images/help/installing-java/multimc/java-in-multimc-4.jpg)
+
+" | markdownify }} </details>
+
+## Verifying if Java is installed {#verifying}
+
+You can check which version of Java may be already installed on your computer by following these instructions.
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:299]:
```suggestion
You can check if a version of Java is already installed on your computer by following these instructions.
```

File: _help/installing-java/index.markdown [hunk 5]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
+Exception in thread "main" java.lang.UnsupportedClassVersionError: net/minecraft/bundler/Main has been compiled by a more recent version of the Java Runtime (class file version 60.0), this version of the Java Runtime only recognizes class file versions up to 55.0
+
+Exception in thread "main" java.lang.UnsupportedClassVersionError: net/minecraft/bundler/Main has been compiled by a more recent version of the Java Runtime (class file version 60.0), this version of the Java Runtime only recognizes class file versions up to 52.0
+```
+```
+Your Java Runtime is incompatible. Please edit your installation to use the bundled Java Runtime.
+Name: Java version mismatch
+```
+```
+(When attempting to launch with MultiMC)
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:336]: We're not dealing with 3rd-party launchers or server software, all of this can go.

Forge can stay, I think, because of its ubiquitousness with normal people (as opposed to enthusiasts or server admins).

File: _help/installing-java/index.markdown [hunk 6]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
+```
+(When attempting to launch Forge)
+java.lang.ClassCastException: class jdk.internal.loader.ClassLoaders$AppClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$AppClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
+```
+
+If you see the above message, you need to *downgrade* to Java 8, because old Forge does not support Java 11+.
+
+*Note: Forge is not officially supported by Minecraft*
+
+```
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:361]: All of this can go, non-Minecraft things (even if they're Minecraft related) are out of scope for this article.

File: _help/installing-java/index.markdown [hunk 7]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
 
-Hold your Windows key (located at the bottom left of your keyboard to the right of the Control key) and press "R".
-You should get a "Run" window open up, type in "cmd" and press "OK."
-* A "Command Prompt" window should open, type in the command: `wmic OS get OSArchitecture`
-* Press enter, and afterwards the second line of the output will tell you if you are using 32-bit or 64-bit.
+Visit the website for [Adoptium Temurin OpenJDK](https://adoptium.net)
 
-#### If you are using 64-bit
-* Go to the [Oracle Java Download page](https://www.java.com/en/download/windows-64bit.jsp) and press the button next to "**Accept License Agreement**" under the "**Java SE Runtime Environment**" header. It should then give you a "**Thank you**" message, once it does look for "jre-8uABC-windows-x64.exe" (where ABC is a two or three digit version code) under the "Download" table - download and install the file.
+Choose the appropriate version, and click **Latest release** to start the download. Note that Temurin versions are the same as Java versions, so "Temurin 17" means "Temurin Java 17". See the above chart to determine which version you need.
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:43]: You need to expand this section to explain how to get older versions of Java/OpenJDK.

File: _help/installing-java/index.markdown [hunk 8]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
+## Do I need Java? Which version do I need?
 
+The Minecraft launcher includes the version(s) of Java required to run the game, and automatically keeps them up to date. This includes the Windows 7 and Windows 8.1 version of the launcher.
 
+If you are using the official launcher, you normally do not have to manually install Java unless you need it for applications other than Minecraft. If you are not are not sure if you are using the correct Minecraft launcher or are getting "This application requires a Java Runtime Environment" when launching Minecraft, then you can download the Minecraft Launcher from [https://minecraft.net/download](https://minecraft.net/download).
 
-# Installing Java<small> - How to install Java</small>
-Note: As of snapshot 21w19a, Java 16 is required to run the game. The launcher should download it automatically, but if you need it to run a server or the launcher doesn't download it automatically, you can download it from [here](https://adoptopenjdk.net/), with installation instructions [here](https://adoptopenjdk.net/installation.html). Be sure to choose OpenJDK 16 from the list.
+The below compatibility chart was last updated 2022-02-20.
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:28]: Get rid of the EOL links on the table, they're confusing for the average person. I'd recommend adding something above the update notice instead, like so:
```suggestion
**Note:** Java versions 6, 8, and 16 no longer receive security updates, and are therefore more risky to use. You should use Java 11 or 17 if you can, as they will receive security updates for years to come.

The below compatibility chart was last updated 2022-02-20.
```
Java 8 EOLs on March 31st, so I also added that to the note.

**Review by Southpaw1496 — CHANGES_REQUESTED** [review]
Review 2, ~~Electric Boogaloo~~

Feel free to reply to my comments or ping me on Discord with any questions :)

File: _help/installing-java/index.markdown [hunk 9]
```diff
@@ -5,71 +5,367 @@ name: "installing-java"
...
+##### RHEL / CentOS / RockyLinux / Amazon Linux
+
+To install Java 17, 11, 8 respectively, choose one of these commands:
+```
+sudo yum install java-17-openjdk
+sudo yum install java-11-openjdk
+sudo yum install java-1.8.0-openjdk
+```
+
+If your distro does not support Java, the Microsoft OpenJDK or Adoptium downloads can be used on all distros.
```
[the remark this answers is not in the document]
Comment (lordofpipes) [reply _help/installing-java/index.markdown:196]: Distributions and package managers are roughly the same thing, but yeah that might be clearer

File: _help/installing-java/index.markdown (hunk 5 above)
Comment (lordofpipes) [reply _help/installing-java/index.markdown:336]: Hmm, I'm disappointed this won't be considered in-scope since excluding it limits the usefulness of the article, and excluding the symptoms limits search engine visibility of the article. Most people are installing Java manually because they're trying to get third-party tools or server software working, since if they were using vanilla it would already be handled

Comment (Southpaw1496) [reply _help/installing-java/index.markdown]: I think that, first and foremost, this site is a tool for MCCS. In general, we only support official things, and while I agree that including additional information about 3rd-party launchers would increase its usefulness outside of supporting official things, it would just increase confusion for users not trying to use those tools, and therefore make it less useful for us specifically.

Comment (Southpaw1496) [reply _help/installing-java/index.markdown]: On top of this, we shouldn't be "supporting" (for want of a better term) 3rd-party tools on the site that we won't support on Discord, that's inconsistent.

File: _help/installing-java/index.markdown (hunk 9 above)
[the remark this answers is not in the document]
Comment (Southpaw1496) [reply _help/installing-java/index.markdown]: Aren't distributions different versions, like Ubuntu, Fedora, Debian etc.

**Comment by lordofpipes** [comment]
> Are you ready for another review or do you still have work to do?

Go for it

File: _help/installing-java/index.markdown [hunk 10]
```diff
@@ -5,71 +5,297 @@ name: "installing-java"
...
+```
+/usr/lib/jvm/java-17-openjdk/bin/java
+/usr/lib/jvm/java-11-openjdk/bin/java
+/usr/lib/jvm/java-8-openjdk/bin/java
+/usr/lib/jvm/java-17-openjdk-amd64/jre/bin/java
+/usr/lib/jvm/java-11-openjdk-amd64/jre/bin/java
+/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java
+```
+
+## Verifying if Java is installed {#verifying}
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:253]: Move this above the tutorial section.

**Review by Southpaw1496 — CHANGES_REQUESTED** [review]
Review 3

If there's anything you are unsure of/disagree with, feel free to discuss it with me before commiting anything :)

**Comment by lordofpipes** [comment]
Looks good, I've done all these changes, fixed a few typos. Thank you!

File: _help/installing-java/index.markdown [hunk 11]
```diff
@@ -5,71 +5,275 @@ name: "installing-java"
...
-- Hold your Windows key (located at the bottom left of your keyboard to the right of the Control key) and press "R".
-You should get a "Run" window open up, type in "cmd" and press "OK.", a "Command Prompt" window should open, type in the command: `java -version`<br>
-Hit Enter and the output should contain `java version "(java version here)"` if Java is installed. If you instead got `'Can't recognise 'java' as an internal or external command, or batch script.` then Java is not installed properly; see the Windows section below for how to install Java.
+Quick links:
 
-### Mac OSX
-- In spotlight, search "Terminal" and open the application.
-Enter in Terminal: `java -version`
-- If the output does not contain java version "[java version here]" then Java is not installed on your computer; see the Mac OSX section below for instructions on how to install Java.
+- [Do I need Java? Which version do I need?](#which-java)
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:17]: You need to change this to reflect the new order of the sections.

**Review by Southpaw1496 — CHANGES_REQUESTED** [review]
Looks excellent, just one more thing that I noticed

**Comment by lordofpipes** [comment]
Fixed

File: _help/installing-java/index.markdown [hunk 12]
```diff
@@ -5,71 +5,275 @@ name: "installing-java"
...
+
+![Terminal with Java alternatives command executed](/static/images/help/installing-java/linux/install-java-4.jpg)
+
+" | markdownify }} </details>
+
+## Next steps
+
+**Additional info:** [Changing the default Java version used by the launcher](/help/guides/changing-java-version/)
+
+**Additional info:** [Verifying if Java is installed properly](#verifying)
```
Comment (Southpaw1496) [inline _help/installing-java/index.markdown:247]: Move this above the Changing launcher. They should verify they have installed Java correctly before trying to use it for anything.

**Review by Southpaw1496 — CHANGES_REQUESTED** [review]
We're getting so close!

**Review by Southpaw1496 — APPROVED** [review]

**Comment by lordofpipes** [comment]
#136 was merged so I went ahead and added platform detection to this new tutorial to keep the pull request up to date - fffa87e

**Review by LordRalex — APPROVED** [review]
